nexus/src/app/background/blueprint_execution.rs

```diff
@@ -0,0 +1,407 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Background task for realizing a plan blueprint
+
+use super::common::BackgroundTask;
+use anyhow::Context;
+use futures::future::BoxFuture;
+use futures::stream;
+use futures::FutureExt;
+use futures::StreamExt;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::lookup::LookupPath;
+use nexus_db_queries::db::DataStore;
+use nexus_types::deployment::{Blueprint, OmicronZonesConfig};
+use serde_json::json;
+use sled_agent_client::Client as SledAgentClient;
+use slog::Logger;
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use tokio::sync::watch;
+use uuid::Uuid;
+
+/// Background task that takes a [`Blueprint`] and realizes the change to
+/// the state of the system based on the `Blueprint`.
+pub struct BlueprintExecutor {
+    datastore: Arc<DataStore>,
+    rx_blueprint: watch::Receiver<Option<Arc<Blueprint>>>,
```

Comment:
I think it's pretty important that we have a way to turn this thing off at runtime if we find the system is doing something harmful.  I believe we already have an `enabled` bool on the BlueprintTarget for this purpose.   So I think maybe this thing should accept `(BlueprintTarget, BlueprintTarget)` (or something similar, instead of just `Blueprint`).  Then this task could just do nothing if `!target.enabled`.

Reply:
> (I assume the suggestion of `(BlueprintTarget, BlueprintTarget)` is a typo and should've been `(Blueprint, BlueprintTarget)`.)
> 
I also assumed this, but reversed the types in the tuple ;)

> Just thinking out loud here, this may dumb: Should we squish `BlueprintTarget` into `Blueprint` somehow, or possibly add a new type for that combination? The db method for reading the current target also returns the 2-tuple, which felt a little awkward. I _think_ the only thing from `BlueprintTarget` that anything other than a debugging human will care about is whether it's enabled, right? Some options in no particular order:
> 
> * Add an `enabled` field to `Blueprint` (I think this doesn't make sense, because what does this field mean if the blueprint isn't the target? but maybe there's some way to make it make sense)
> * Add a `blueprint` field to `BlueprintTarget`
> * Add a struct that has `Blueprint` and `BlueprintTarget` as fields
> * Add `EnabledBlueprint` / `DisabledBlueprint` newtypes around `Blueprint`, which can be created from a `BlueprintTarget` and its `Blueprint`

I think dealing with tuples is a pain in the ass, but I don't like the idea of squishing these in the DB. Ideally the Blueprint is immutable, and the pointer to it (the target) is mutable. This would change that. The newtype idea could work though.

Reply:
Ah, I didn't mean squishing in the db itself, just in the types / return values from the db query. But yeah I think the newtype is maybe my preference, although maybe it should be an enum? Something like

```rust
enum CurrentTargetBlueprint {
    Enabled(Blueprint),
    Disabled(Blueprint),
}
```

?

Reply:
Hm, this loses the extra debugging info that's present in `BlueprintTarget`. Maybe I'm now leaning toward squishing the tuple into a struct just to give it a name?

```rust
struct CurrentTargetBlueprint {
    blueprint: Blueprint,
    metadata: BlueprintTarget,
}
```

or something.

Reply:
Deleted my last comment, because it made no sense. lol.

Reply:
I implemented this using the current tuple version in [9d95fd8](https://github.com/oxidecomputer/omicron/pull/4891/commits/9d95fd84dc23eee190c1924cf57ff7574617acb7)

I'm happy to change to use to a new type if we agree, although it can probably wait for a follow up.